components/log4j/org.wso2.carbon.analytics.shared.data.agents.log4j/src/main/java/org/wso2/carbon/analytics/shared/data/agents/log4j/appender/LogEventAppender.java

```diff
@@ -430,8 +433,18 @@ private void publishLogEvent(TenantDomainAwareLoggingEvent event) throws ParseEx
             if (event.getThrowableInformation() != null) {
                 arbitraryDataMap.put(columns[8], stacktrace);
             }
-            Event laEvent = new Event(streamDef, date.getTime(), null, null, new String[]{tenantDomain}, arbitraryDataMap);
-            dataPublisher.publish(laEvent);
+            Event logEvent = new Event(streamDef, date.getTime(), null, null, new String[]{tenantDomain},
+                    arbitraryDataMap);
+            if (dataPublisher != null) {
+                dataPublisher.publish(logEvent);
+            } else {
+                publisherInitializer();
```

Comment:
In the concurrent environment there is a possibility that many data publishers getting created, because there is no precautions taken to have only single datapublisher instance. This is highly likely to happen because the data publisher instance is being created during the attempt to publishing the event, not during bundle/class loading time. Either we have to initialize the data publisher during the load time, or else we atleast need to make the initializer method call as synchrnonized, and check for the object instantiation.